README.md
# React Responsive

[![Build Status](https://travis-ci.org/dvaJi/react-responsive.svg?branch=master)](https://travis-ci.org/dvaJi/react-responsive) [![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)

A simple React library to react to responsive events using the new context api.

## Installation

```
With NPM:
npm i --save @dvaji/react-responsive

With Yarn:
yarn add @dvaji/react-responsive
```

### Example

Include the provider for giving the context

```jsx
import { ResponsiveProvider } from '@dvaji/react-responsive'

ReactDOM.render(
  <ResponsiveProvider>
    <App />
  </ResponsiveProvider>,
  document.getElementById('root')
)
```

Use the consumer in your components

```jsx
import { ResponsiveConsumer } from '@dvaji/react-responsive'
;() => (
  <ResponsiveConsumer>
    {status => (
      <div>
        {status.desktop && <p>This only shows in Desktop</p>}
        <Button size={status.mobile ? 'small' : 'large'}>Click me</Button>
      </div>
    )}
  </ResponsiveConsumer>
)
```

### Default sizes and configuration

Default sizes:

- **mobile**: up to 767px
- **tablet**: from 768px to 1023px
- **desktop**: from 1024px

you can set any ranges you want like this

```jsx
import { ResponsiveProvider } from '@dvaji/react-responsive'

// An object with your custom ranges
const sizes = {
  small: [-Infinity, 500],
  medium: [501, 1000],
  large: [1001, +Infinity]
}

ReactDOM.render(
  <ResponsiveProvider sizes={sizes}>
    <App />
  </ResponsiveProvider>,
  document.getElementById('root')
)
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.0.2"></a>

## [0.0.2](https://github.com/dvaJi/react-responsive/compare/v0.0.1...v0.0.2) (2018-09-02)

<a name="0.0.1"></a>

## 0.0.1 (2018-09-02)


src/__tests__/Responsive.test.js
import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { ResponsiveProvider, ResponsiveConsumer } from '../index'

Enzyme.configure({ adapter: new Adapter() })

beforeEach(() => {
  jest.resetModules()
  global.window.resizeTo(1024)
})

it('should render without throwing an error', () => {
  const wrapper = shallow(
    <ResponsiveProvider>
      <div>
        <ResponsiveConsumer>
          {status => (
            <div>
              {status.desktop && <p>This only shows in Desktop</p>}
              <button size={status.mobile ? 'small' : 'large'}>Click me</button>
            </div>
          )}
        </ResponsiveConsumer>
      </div>
    </ResponsiveProvider>
  )
  expect(wrapper).toBeTruthy()
  wrapper.unmount()
})

it('should receive custom ranges without throwing an error', () => {
  const sizes = {
    small: [-Infinity, 500],
    medium: [501, 1000],
    large: [1001, +Infinity]
  }
  const wrapper = mount(
    <ResponsiveProvider sizes={sizes}>
      <div>
        <ResponsiveConsumer>
          {status => (
            <div>
              {status.large && <p>This only shows in Large size</p>}
              <button size={status.small ? 'small' : 'large'}>Click me</button>
            </div>
          )}
        </ResponsiveConsumer>
      </div>
    </ResponsiveProvider>
  )
  expect(wrapper).toBeTruthy()
  wrapper.unmount()
})

it('should render the text according to the window size without throwing an error', () => {
  const wrapper = mount(
    <ResponsiveProvider>
      <div>
        <ResponsiveConsumer>
          {status => (
            <div>
              {status.mobile && <p id="is-mobile">This only shows in Mobile</p>}
              {status.tablet && <p id="is-tablet">This only shows in Tablet</p>}
              {status.desktop && (
                <p id="is-desktop">This only shows in Desktop</p>
              )}
            </div>
          )}
        </ResponsiveConsumer>
      </div>
    </ResponsiveProvider>
  )
  expect(wrapper).toBeTruthy()
  expect(wrapper.find('#is-desktop').render()[0].children[0].data).toEqual(
    'This only shows in Desktop'
  )

  global.window.resizeTo(798)
  wrapper.update()
  expect(wrapper.find('#is-tablet').render()[0].children[0].data).toEqual(
    'This only shows in Tablet'
  )

  global.window.resizeTo(765)
  wrapper.update()
  expect(wrapper.find('#is-mobile').render()[0].children[0].data).toEqual(
    'This only shows in Mobile'
  )

  wrapper.unmount()
})

it('should render the text according to the window size with custom ranges', () => {
  const sizes = {
    small: [-Infinity, 500],
    medium: [501, 1000],
    large: [1001, +Infinity]
  }

  const wrapper = mount(
    <ResponsiveProvider sizes={sizes}>
      <div>
        <ResponsiveConsumer>
          {status => (
            <div>
              {status.small && <p id="is-small">Small</p>}
              {status.medium && <p id="is-medium">Medium</p>}
              {status.large && <p id="is-large">Large</p>}
            </div>
          )}
        </ResponsiveConsumer>
      </div>
    </ResponsiveProvider>
  )

  expect(wrapper).toBeTruthy()
  expect(wrapper.find('#is-large').render()[0].children[0].data).toEqual(
    'Large'
  )

  global.window.resizeTo(798)
  wrapper.update()
  expect(wrapper.find('#is-medium').render()[0].children[0].data).toEqual(
    'Medium'
  )

  global.window.resizeTo(480)
  wrapper.update()
  expect(wrapper.find('#is-small').render()[0].children[0].data).toEqual(
    'Small'
  )

  wrapper.unmount()
})


setupFile.js
import { JSDOM } from 'jsdom'

const jsdom = new JSDOM('<!doctype html><html><body></body></html>')
const { window } = jsdom

function copyProps(src, target) {
  const props = Object.getOwnPropertyNames(src)
    .filter(prop => typeof target[prop] === 'undefined')
    .reduce(
      (result, prop) => ({
        ...result,
        [prop]: Object.getOwnPropertyDescriptor(src, prop)
      }),
      {}
    )
  Object.defineProperties(target, props)
}

global.window = window
global.window.resizeTo = (width, height) => {
  global.window.innerWidth = width || global.window.innerWidth
  global.window.innerHeight = height || global.window.innerHeight
  global.window.dispatchEvent(new Event('resize')) // eslint-disable-line no-undef
}
global.document = window.document
global.navigator = {
  userAgent: 'node.js'
}
copyProps(window, global)


rollup.config.js
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs'
import babel from 'rollup-plugin-babel'
import filesize from 'rollup-plugin-filesize'

const name = 'react-responsive'
const globals = {
  react: 'React',
  'prop-types': 'PropTypes'
}

export default {
  input: 'src/index.js',
  output: [
    {
      name,
      globals,
      format: 'umd',
      file: 'index.js'
    },
    {
      name,
      globals,
      format: 'es',
      file: 'index.es.js'
    }
  ],
  external: ['prop-types', 'react'],
  plugins: [
    resolve({
      jsnext: true,
      main: true
    }),
    babel({
      babelrc: false,
      presets: [
        [
          'env',
          {
            debug: true,
            targets: { browsers: ['> 1%', 'last 2 versions', 'ie > 9'] },
            modules: false
          }
        ],
        'stage-1',
        'react'
      ],
      plugins: ['external-helpers'],
      exclude: 'node_modules/**'
    }),
    commonjs(),
    filesize()
  ]
}


src/components/Responsive.js
import React, { PureComponent } from 'react'

const providerMap = (object, func) =>
  Object.assign(
    {},
    ...Object.keys(object).map(key => ({ [key]: func(object[key]) }))
  )

const defaultSizes = {
  mobile: [-Infinity, 767],
  tablet: [768, 1023],
  desktop: [1024, +Infinity]
}

export const ResponsiveContext = React.createContext({})

export class ResponsiveProviderBase extends PureComponent {
  constructor(props) {
    super(props)
    this.state = this.actualResponsiveState()
  }

  actualResponsiveState() {
    const actualWidth = this.width()
    const detectSize = size => actualWidth >= size[0] && actualWidth <= size[1]
    return providerMap(this.props.sizes, detectSize)
  }

  render() {
    return (
      <ResponsiveContext.Provider value={this.state}>
        {this.props.children}
      </ResponsiveContext.Provider>
    )
  }

  onWindowResize = () => this.setState(this.actualResponsiveState())
}

ResponsiveProviderBase.defaultProps = {
  sizes: defaultSizes
}


index.es.js
import React, { PureComponent } from 'react';

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var defineProperty = function (obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
};

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var toConsumableArray = function (arr) {
  if (Array.isArray(arr)) {
    for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) arr2[i] = arr[i];

    return arr2;
  } else {
    return Array.from(arr);
  }
};

var providerMap = function providerMap(object, func) {
  return Object.assign.apply(Object, [{}].concat(toConsumableArray(Object.keys(object).map(function (key) {
    return defineProperty({}, key, func(object[key]));
  }))));
};

var defaultSizes = {
  mobile: [-Infinity, 767],
  tablet: [768, 1023],
  desktop: [1024, +Infinity]
};

var ResponsiveContext = React.createContext({});

var ResponsiveProviderBase = function (_PureComponent) {
  inherits(ResponsiveProviderBase, _PureComponent);

  function ResponsiveProviderBase(props) {
    classCallCheck(this, ResponsiveProviderBase);

    var _this = possibleConstructorReturn(this, (ResponsiveProviderBase.__proto__ || Object.getPrototypeOf(ResponsiveProviderBase)).call(this, props));

    _this.onWindowResize = function () {
      return _this.setState(_this.actualResponsiveState());
    };

    _this.state = _this.actualResponsiveState();
    return _this;
  }

  createClass(ResponsiveProviderBase, [{
    key: 'actualResponsiveState',
    value: function actualResponsiveState() {
      var actualWidth = this.width();
      var detectSize = function detectSize(size) {
        return actualWidth >= size[0] && actualWidth <= size[1];
      };
      return providerMap(this.props.sizes, detectSize);
    }
  }, {
    key: 'render',
    value: function render() {
      return React.createElement(
        ResponsiveContext.Provider,
        { value: this.state },
        this.props.children
      );
    }
  }]);
  return ResponsiveProviderBase;
}(PureComponent);

ResponsiveProviderBase.defaultProps = {
  sizes: defaultSizes
};

var ResponsiveConsumer = ResponsiveContext;

var ResponsiveProvider = function (_ResponsiveProviderBa) {
  inherits(ResponsiveProvider, _ResponsiveProviderBa);

  function ResponsiveProvider() {
    classCallCheck(this, ResponsiveProvider);
    return possibleConstructorReturn(this, (ResponsiveProvider.__proto__ || Object.getPrototypeOf(ResponsiveProvider)).apply(this, arguments));
  }

  createClass(ResponsiveProvider, [{
    key: 'width',
    value: function width() {
      return window.innerWidth;
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener('resize', this.onWindowResize, false);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      window.removeEventListener('resize', this.onWindowResize);
    }
  }]);
  return ResponsiveProvider;
}(ResponsiveProviderBase);

export { ResponsiveConsumer, ResponsiveProvider };
